maplibre/src/coords.rs

```diff
@@ -82,36 +78,45 @@ impl fmt::Debug for Quadkey {
     Default,
     Serialize,
     Deserialize,
-    Pod,
     Zeroable,
 )]
 #[repr(C)]
 pub struct ZoomLevel(u8);
 
 impl ZoomLevel {
     pub const fn new(z: u8) -> Self {
-        ZoomLevel(z)
+        if z > MAX_ZOOM as u8 {
+            Self(MAX_ZOOM as u8)
+        } else {
+            Self(z)
+        }
     }
     pub fn is_root(self) -> bool {
         self.0 == 0
     }
-}
 
-impl std::ops::Add<u8> for ZoomLevel {
-    type Output = ZoomLevel;
+    pub(crate) fn max_tile_coord(&self) -> u32 {
+        (1 << self.0) - 1
+    }
 
-    fn add(self, rhs: u8) -> Self::Output {
-        let zoom_level = self.0.checked_add(rhs).expect("zoom level overflowed");
-        ZoomLevel(zoom_level)
+    pub const fn zoom_in(&self) -> Option<ZoomLevel> {
+        if self.0 == MAX_ZOOM {
+            None
+        } else {
+            Some(Self(self.0 + 1))
+        }
     }
-}
 
-impl std::ops::Sub<u8> for ZoomLevel {
-    type Output = ZoomLevel;
+    pub const fn zoom_out(&self) -> Option<ZoomLevel> {
+        if self.0 == 0 {
+            None
+        } else {
+            Some(Self(self.0 - 1))
+        }
+    }
 
-    fn sub(self, rhs: u8) -> Self::Output {
-        let zoom_level = self.0.checked_sub(rhs).expect("zoom level underflowed");
-        ZoomLevel(zoom_level)
+    pub(crate) fn as_u8(&self) -> u8 {
```

Comment:
```suggestion
    pub fn as_u8(&self) -> u8 {
```

So far I'm not using pub(crate) anywhere.

Reply:
Should we implement Deref instead of as_u8?

Reply:
Imo `Deref`` counteracts the new type idea. I like the explicitness more.

I used `pub(crate)` because this feels very "internal" to me but I guess the question what will be stable API and what wont be is not really relevent yet.